Add Parallel WaveGan and example

iclementine opened the pull request:
1. Add Parallel WaveGAN model;
2. Add a general purpose dataset;
3. Add a complete example for training parallel wavegan on Baker dataset;
4. Add some utilities for training.

Commit:
add kbest

Commit:
fix typos

Commit:
remove unnecessary imports

Commit:
fix del fn

Commit:
fix typos

Commit:
add k-latest

Commit:
STFT and MelScale: register filters as buffer.

Commit:
add parallel wavegan model

Commit:
add AudioDataset and MelDataset

Commit:
Merge branch 'develop' of https://github.com/PaddlePaddle/Parakeet into pwg

Commit:
add h5py utility to interact with numpy

Commit:
add AudioMelDataset for training vocoders.

zh794390558 commented on parakeet/datasets/audio_dataset.py:
```diff
@@ -0,0 +1,133 @@
+
+        # filter by threshold
+        if audio_length_threshold is not None:
+            audio_lengths = [audio_load_fn(f).shape[0] for f in audio_files]
```
这种方式会很耗费io的，有一个list文件，需要的数据提前处理好会好很多。

iclementine replied:
提前把音频的帧数和 sample 个数记录下来。是的。在预处理的时候全部记录下来会更好。就不用在 filter 数据的时候完全加载了。

zh794390558 commented on parakeet/datasets/audio_dataset.py:
```diff
@@ -0,0 +1,133 @@
+        The root of the dataset.
+    audio_pattern : str
+        A pattern to recursively find all audio files, by default "*-wave.npy"
+    audio_length_threshold : int, optional
```
min_audio_length

iclementine replied:
okay，这些我都统一改一下

Commit:
checkpoint utility: add support for str in addition to Path object

zh794390558 commented on parakeet/datasets/audio_dataset.py:
```diff
@@ -0,0 +1,133 @@
+        self.audio_files = audio_files
+        self.audio_load_fn = audio_load_fn
+        self.return_utt_id = return_utt_id
+        # TODO(chenfeiyu): better strategy to get utterance id
```
可以参考识别的manifest风格。

iclementine replied:
好的。我参考一下怎么规范这个文件的格式和位置。如果方便，直接让它成为 preprocess 结果中的必须包含文件。这样就可以在读取的时候使用了。

zh794390558 commented on parakeet/datasets/audio_dataset.py:
```diff
@@ -0,0 +1,133 @@
+        else:
+            self.utt_ids = [f.stem for f in audio_files]
+        self.use_cache = use_cache
+        if use_cache:
```
这块儿的目的是啥？

iclementine replied:
可以懒惰地把读取过的 example 放在内存中。下次读取的时候不用从文件系统加载。

zh794390558 commented on parakeet/datasets/audio_dataset.py:
```diff
@@ -0,0 +1,133 @@
+        audio = self.audio_load_fn(self.audio_files[idx])
+
+        if self.return_utt_id:
+            items = utt_id, audio
```
可以把这个当做默认模式。可以考虑在这里不做load_audio的操作，放在collactor里做。collactor基本能满足所有的场景需求，可以batch的处理audio。

iclementine replied:
把耗时的操作放在 __getitem__ 是我们目前的默认做法。因为根据此前的实验，这么做能更高效地读取大 batch.

zh794390558 commented on parakeet/datasets/audio_mel_dataset.py:
```diff
@@ -0,0 +1,161 @@
+import logging
+
+
+class AudioMelDataset(Dataset):
```
不如做一个dataset，读取utt, audio_path, mel_path, linear_path这样来的舒服。

iclementine replied:
可以设计一下，之后加一个

iclementine replied:
已经添加了 DataTable， 类似 DataFrame 的设计。

zh794390558 commented on parakeet/datasets/audio_mel_dataset.py:
```diff
@@ -0,0 +1,161 @@
+        mel_files = sorted(list(root_dir.rglob(mel_pattern)))
+
+        # filter by threshold
+        if audio_length_threshold is not None:
```
min_audio_length

iclementine replied:
这个已经去掉了

zh794390558 commented on parakeet/datasets/audio_mel_dataset.py:
```diff
@@ -0,0 +1,161 @@
+                    f"({len(audio_files)} -> {len(idxs)}).")
+            audio_files = [audio_files[idx] for idx in idxs]
+            mel_files = [mel_files[idx] for idx in idxs]
+        if mel_length_threshold is not None:
```
同上。

iclementine replied:
这个已经去掉了

zh794390558 commented on parakeet/datasets/mel_dataset.py:
```diff
@@ -0,0 +1,132 @@
+import logging
+
+
+class MelDataset(Dataset):
```
同上

iclementine replied:
已经去除。

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,438 @@
+        self.y_scale = y_scale
+        self.mode = mode
+
+    def forward(self, x):
```
写模型加上输入x的shape信息。

iclementine replied:
okay, 会加的。

Commit:
remove task-specific Datasets and add a general purpose DataTable

Commit:
move docstring to class

Commit:
fix file names

Commit:
WIP: add sample code for parallel wavegan

Commit:
fix imports

Commit:
WIP: training setup done

Commit:
WIP: pwg training works

Commit:
WIP: add some trainig info

zh794390558 commented on parakeet/datasets/data_table.py:
```diff
@@ -0,0 +1,151 @@
+            self._initialize_cache()
+
+    def _initialize_cache(self):
+        self.manager = Manager()
```
Manager 看不出这个类是做什么的。

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,678 @@
+        self.use_causal_conv = use_causal_conv
+        self.up_layers = nn.LayerList()
+        for scale in upsample_scales:
+            stretch = Stretch2D(scale, 1, interpolate_mode)
```
带上参数名

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,678 @@
+        self.aux_context_window = aux_context_window
+        self.use_causal_conv = use_causal_conv and aux_context_window > 0
+        kernel_size = aux_context_window + 1 if use_causal_conv else 2 * aux_context_window + 1
+        self.conv_in = nn.Conv1D(
```
use_causal_conv时没有padding？

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,678 @@
+        c = c.unsqueeze(1)
+        for f in self.up_layers:
+            if self.use_causal_conv and isinstance(f, nn.Conv2D):
+                c = f(c)[:, :, :, c.shape[-1]]
```
建议不用c.shape

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,678 @@
+
+
+class Stretch2D(nn.Layer):
+    def __init__(self, w_scale: int, h_scale: int, mode: str="nearest"):
```
w, h 参数换位置。

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,678 @@
+        x : Tensor
+            Shape (N, C_res, T), the input features.
+        c : Tensor
+            Shape (N, C_aux, T), he auxiliary input.
```
The

zh794390558 replied:
auxiliary 最好换成condition 或者 local condition，或者注释中说明吧。

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,678 @@
+        x = paddle.tanh(a) * F.sigmoid(b)
+
+        skip = self.conv1x1_skip(x)
+        res = (self.conv1x1_out(x) + x_input) * math.sqrt(0.5)
```
math.sqrt 替换成 paddle

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,678 @@
+        assert layers % stacks == 0
+        layers_per_stack = layers // stacks
+
+        self.first_conv = nn.Conv1D(
```
first_1x1conv, 为了对齐名字就不用修改了。

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,678 @@
+        for f in self.conv_layers:
+            x, s = f(x, c)
+            skips += s
+        skips *= math.sqrt(1.0 / len(self.conv_layers))
```
换成paddle

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,678 @@
+
+        if c is not None:
+            c = paddle.transpose(c, [1, 0]).unsqueeze(0)  # pseudo batch
+            c = nn.Pad1D(self.aux_context_window, mode='edge')(c)
```
Pad1D的文档没有更新！

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,678 @@
+        for f in self.conv_layers:
+            x, h = f(x, None)
+            skip += h
+        skip *= math.sqrt(1 / len(self.conv_layers))
```
替换paddle

zh794390558 commented on parakeet/models/parallel_wavegan.py:
```diff
@@ -0,0 +1,678 @@
+        x = self.last_conv_layers(x)
+        return x
+
+    def apply_weight_norm(self):
```
添加WeightNormLayer？

iclementine replied:
这个不需要。因为有 weight norm hook, 而且每个模型模型可能涉及的加了 weight norm 的 Layer 类型不同，之所以把它弄成类的方法是为了这个类自定义的需要。

如果需要不做过滤地递归去除，也可以用 utils.layer_tools.recursively_remove_weight_norm

zh794390558 commented on parakeet/modules/audio.py:
```diff
@@ -195,44 +193,42 @@ def forward(self, x):
         """
         power = self.power(x)
-        magnitude = paddle.sqrt(power)
+        magnitude = paddle.sqrt(power)  # TODO(chenfeiyu): maybe clipping
```
add eps

iclementine replied:
应该不需要了，sqrt(0) 不会有问题。有问题的是 log(0), 只需要在 log 那里处理就好。

zh794390558 commented on parakeet/modules/stft_loss.py:
```diff
@@ -0,0 +1,138 @@
+        Returns:
+            Tensor: Log STFT magnitude loss value.
+        """
+        return F.l1_loss(paddle.log(y_mag), paddle.log(x_mag))
```
add eps

iclementine replied:
已经修改

iclementine replied:
已经修改

zh794390558 commented on parakeet/training/checkpoint.py:
```diff
@@ -12,7 +12,8 @@
 # limitations under the License.
 
-from typing import Callable, Mapping, List
+from typing import Callable, Mapping, List, Union
```
one import per line

iclementine replied:
能说说理由？
我不觉得从一个模块里导入多个类的时候这么写有什么问题。
PEP 8 也没有如此的要求。

zh794390558 commented on parakeet/utils/h5_utils.py:
```diff
@@ -0,0 +1,104 @@
+            f"There is no such a data in hdf5 file. ({dataset_name})")
+        sys.exit(1)
+
+    hdf5_data = hdf5_file[dataset_name][()]  # a special syntax of h5py
```
An empty tuple (()) to retrieve all data or scalar data

iclementine replied:
已经修改

zh794390558 commented on examples/parallelwave_gan/baker/batch_fn.py:
```diff
@@ -0,0 +1,108 @@
+        """
+        # check length
+        examples = [
+            self._adjust_length(b['wave_path'], b['feats_path'])
```
wav_path -> wav

iclementine replied:
已经修改为 wave. 一个特征变换前后可能性质会变，但是语义保持一致。
path 也是 wave, array 也是 wave

zh794390558 commented on examples/parallelwave_gan/baker/preprocess.py:
```diff
@@ -0,0 +1,280 @@
+    fmax = sr / 2 if fmax is None else fmax
+    mel_basis = librosa.filters.mel(sr, n_fft, n_mels, fmin, fmax)
+
+    return np.log10(np.maximum(eps, np.dot(mel_basis, spc)))
```
log10 -> log

iclementine replied:
这个暂不修改。等完整训练完一版再作修改。

Commit:
add profiling

Commit:
add profiling

Commit:
use paddle's profiler

Commit:
add unit testings back

Commit:
add some profiling to unittesting

Commit:
add code to compute statistics

Commit:
move synchronize up before timer starts

Commit:
1. add compute_statistics and normalize;
2. use jsonlines to read and write metadata by default;
3. use threadpool to replace processpool in preprocessing cause it is faster.

Commit:
add docstring to pwg

Commit:
add docstrings to pwg

Commit:
1. change default data layout to channel last in preprocessing;
2. add Summary and DictSummary for aggrelation of evaluation losses;
3. add unittest for report ans scope.

Commit:
add snapshot and visualizer

Commit:
add numerical testing for PWGGenerator

zh794390558 commented on examples/parallelwave_gan/baker/preprocess.py:
```diff
@@ -190,8 +190,9 @@ def process_sentences(config,
-        json.dump(results, f)
+    with jsonlines.open(output_dir / "metadata.jsonl", 'w') as writer:
+        for item in results:
+            writer.write(item)
```
json.dumps(item) + '\n'

iclementine replied:
已经改用 jsonlines 来处理，因为这个格式就是 jsonlines.

Commit:
add profiling tool and a config for parallel training

Commit:
add expanuser for preprocess, updater backward testing for pwg

Commit:
add synthesis script for pwg

Commit:
make a better snapshot extension

Commit:
add better traininig utility code

Commit:
1. better error handling;
2. use absolute path in snapshot records;
3. visualdl takes a logger as init argument rather than creating one.

Commit:
refined training module

Commit:
fix priority in example/parallelwavegan

Commit:
auto choose device for inference

Commit:
add --device to cli argument, use gpu by default

Commit:
delete config for paralle training

zh794390558 commented on examples/parallelwave_gan/baker/batch_fn.py:
```diff
@@ -0,0 +1,110 @@
+            batch_max_steps=20480,
+            hop_size=256,
+            aux_context_window=0, ):
+        """Initialize customized collater for PyTorch DataLoader.
```
PyTorch DataLoader

iclementine replied:
已经修改

zh794390558 commented on examples/parallelwave_gan/baker/conf/default.yaml:
```diff
@@ -0,0 +1,129 @@
+# global_gain_scale: 1.0   # Will be multiplied to all of waveform.
+trim_silence: false      # Whether to trim the start and end of silence.
+top_db: 60 # Need to tune carefully if the recording is not good.
+trim_frame_length: 2048    # Frame size in trimming.
```
sample or frame?

iclementine replied:
samples

iclementine replied:
添加了一些说明，但是这种默认都用 num_samples 来说明。因为它说的其实就是 frame 的大小，那肯定是用 num_samples 来衡量的

zh794390558 commented on examples/parallelwave_gan/baker/compute_statistics.py:
```diff
@@ -0,0 +1,108 @@
+    parser.add_argument(
+        "--metadata", type=str, help="json file with id and file paths ")
+    parser.add_argument(
+        "--field-name", type=str, help="json file with id and file paths ")
```
help 不匹配

iclementine replied:
已经修改。

zh794390558 commented on examples/parallelwave_gan/baker/preprocess.py:
```diff
@@ -0,0 +1,281 @@
+    return np.log10(np.maximum(eps, np.dot(mel_basis, spc)))
+
+
+def process_sentence(config: Dict[str, Any],
```
抽出 trim_silence

iclementine replied:
这貌似已经没有什么好抽的了，它就是 Librosa.effects.trim

zh794390558 commented on examples/parallelwave_gan/baker/preprocess.py:
```diff
@@ -0,0 +1,281 @@
+
+    # adjust time to make num_samples == num_frames * hop_length
+    num_frames = logmel.shape[1]
+    y = np.pad(y, (0, config.n_fft), mode="reflect")
```
建议精确计算pad值

iclementine replied:
已经修改。

zh794390558 commented on examples/parallelwave_gan/baker/preprocess.py:
```diff
@@ -0,0 +1,281 @@
+        list((root_dir / "PhoneLabeling").rglob("*.interval")))
+
+    # split data into 3 sections
+    train_wav_files = wav_files[:9800]
```
用宏替代常数

iclementine replied:
已经修改

zh794390558 commented on examples/parallelwave_gan/baker/pwg_updater.py:
```diff
@@ -0,0 +1,188 @@
+        ## Adversarial loss
+        if self.state.iteration > self.discriminator_train_start_steps:
+            with timer() as t:
+                p_ = self.discriminator(wav_)
```
discriminator不计算wav的loss吗？

zh794390558 commented on examples/parallelwave_gan/baker/pwg_updater.py:
```diff
@@ -0,0 +1,188 @@
+        # Disctiminator
+        if self.state.iteration > self.discriminator_train_start_steps:
+            with paddle.no_grad():
+                wav_ = self.generator(noise, mel)
```
这部分是否可以省略，detach后使用？

iclementine replied:
detach 和 no_grad 不一样，deatch 是复制一个节点并且切断梯度回传。但是 no_grad 是在整个正向运行过程中不构建反向图，所以当不需要梯度传播的时候，这样才是正常的做法。

zh794390558 commented on examples/parallelwave_gan/baker/pwg_updater.py:
```diff
@@ -0,0 +1,188 @@
+
+        # Disctiminator
+        with paddle.no_grad():
+            wav_ = self.generator(noise, mel)
```
同样的问题

iclementine replied:
这里可以去掉，减少一次重复计算

zh794390558 commented on examples/parallelwave_gan/baker/synthesize.py:
```diff
@@ -0,0 +1,93 @@
+state_dict = paddle.load(args.checkpoint)
+generator.set_state_dict(state_dict["generator_params"])
+
+generator.remove_weight_norm()
```
不做的话会有什么结果？

iclementine replied:
这么做是为了让推理速度变快，不这么做就会慢一些

Commit:
minor fixes to refine code.

Commit:
avoid duplicated computation in validation, compute adversarial before stft loss.

zh794390558 reviewed (approved):
LGTM
